src/FsAutoComplete/CodeFixes/ExprTypeMismatch.fs

```diff
@@ -0,0 +1,117 @@
+  let visitor =
+    { new SyntaxVisitorBase<bool>() with
+        member _.VisitExpr(path, traverseSynExpr, defaultTraverse, synExpr) =
+          if not (Range.equals synExpr.Range diagnosticRange) then
+            defaultTraverse synExpr
+          else
+            match synExpr with
+            | SynExpr.Const _
+            | SynExpr.Ident _ -> Some false
+            | e -> defaultTraverse e }
+
+  SyntaxTraversal.Traverse(diagnosticRange.Start, tree, visitor)
+  |> Option.defaultValue true
+
+let title = "ExprTypeMismatch Codefix"
+
+let fix (getParseResultsForFile: GetParseResultsForFile) : CodeFix =
+  Run.ifDiagnosticByCode (set [ "1" ]) (fun diagnostic (codeActionParams: CodeActionParams) ->
+    asyncResult {
+      let fileName = codeActionParams.TextDocument.GetFilePath() |> Utils.normalizePath
+      let fcsPos = protocolPosToPos diagnostic.Range.Start
+
+      let! (parseAndCheckResults: ParseAndCheckResults, _line: string, sourceText: IFSACSourceText) =
+        getParseResultsForFile fileName fcsPos
+
+      let diagnosticWithExtendedData =
+        parseAndCheckResults.GetCheckResults.Diagnostics
+        |> Array.tryPick (fun d ->
+          match d.ExtendedData with
+          | Some(:? TypeMismatchDiagnosticExtendedData as data) -> Some(d, data)
+          | _ -> None)
+
+      match diagnosticWithExtendedData with
+      | None -> return []
+      | Some(diagnostic, extendedData) ->
+        let updateReturnType =
+          findReturnType fcsPos parseAndCheckResults.GetParseResults.ParseTree
+          |> Option.map (fun mReturnType ->
+            let currentType = sourceText.GetSubTextFromRange mReturnType
+            let actualType = extendedData.ActualType.Format(extendedData.DisplayContext)
+
+            { SourceDiagnostic = None
+              Title = $"Update %s{currentType} to %s{actualType}"
+              File = codeActionParams.TextDocument
+              Edits =
+                [| { Range = fcsRangeToLsp mReturnType
+                     NewText = actualType } |]
+              Kind = FixKind.Fix })
+          |> Option.toList
+
+        let optionFixes =
+          if diagnostic.Range.StartLine <> diagnostic.Range.EndLine then
+            []
+          elif
+            extendedData.ExpectedType.BasicQualifiedName = "Microsoft.FSharp.Core.option`1"
+            || extendedData.ExpectedType.BasicQualifiedName = "Microsoft.FSharp.Core.voption`1"
+          then
+            let currentExpr = sourceText.GetSubTextFromRange diagnostic.Range
+
+            let isValueOption =
+              extendedData.ExpectedType.BasicQualifiedName = "Microsoft.FSharp.Core.voption`1"
+
+            let wrapIn = if isValueOption then "ValueSome" else "Some"
+            let replaceWithNone = if isValueOption then "ValueNone" else "None"
+
+            let needsParenthesis =
+              needParenthesisWhenWrappedInSome diagnostic.Range parseAndCheckResults.GetParseResults.ParseTree
```

Comment:
Doesn't apply now, but once the next SDK is shipped, I think this can become something like

```fsharp
(diagnostic.Range.Start, parseAndCheckResults.GetAST)
||> ParsedInput.tryNode
|> Option.forall (fun (node, path) ->
    match node with
    | SyntaxNode.SynExpr expr -> SynExpr.shouldBeParenthesizedInContext getSourceLineStr path expr
    | _ -> true)
```

Reply:
Yes, I was hoping to use that one in the next SDK!